File: python/layer.py
import numpy as np
from utils import relu, softmax

class Layer(object):
    def __init__(self, prev, neurons, end):
        self.neurons = neurons
        self.prev = prev
        self.end = end
        
        self.weights = None
        self.bias = None
    
    def initial_parameters(self, weights, bias):
        if weights is None:
            self.weights = np.random.randn(self.neurons, self.prev) * np.sqrt(1 / self.prev)
        else:
            self.weights = weights;
            
        if bias is None:
            self.bias = np.zeros((self.neurons, 1))
        else:
            self.bias = bias
            
    def forward(self, input):
        if self.weights is None:
            raise Exception("Weights no initialization!")
        
        z = self.weights.dot(input) + self.bias
        
        if self.end:
            return softmax(z)
        else:
            return relu(z)
        

File: javascript/utils/layer.js
const math = require('../math.min');
const { relu, softmax } = require('./activation');
const { broadcastAdd } = require('./utils');

class Layer {
	constructor(...args) {
		this.neurons = args[1];
		this.prev = args[0];
		this.end = args[2];
		this.weights = null;
		this.bias = null;
	}

	initialParameters(weights, bias) {
		if (weights && bias) {
			this.weights = weights;
			this.bias = bias;
		} else {
			this.weights = math.multiply(math.random([ this.neurons, this.prev ], -1, 1), 0.01);
			this.bias = Array(this.neurons).fill(0);
		}
	}

	// wx + b
	forward(input) {
		if (this.weights === null) {
			throw new Error('no initialization of weights');
		}

		let z = broadcastAdd(math.multiply(this.weights, input), this.bias);

		if (this.end) {
			return softmax(z);
		} else {
			return relu(z);
		}
	}
}

module.exports = Layer;
